src/main/java/com/microsoft/sqlserver/jdbc/KeyVaultCredential.java

```diff
@@ -5,86 +5,118 @@
-    String clientId = null;
-    String clientKey = null;
-    String accessToken = null;
+@Immutable
+class KeyVaultCredential implements TokenCredential {
+    private final ClientLogger logger = new ClientLogger(KeyVaultCredential.class);
+    private final String clientId;
+    private final String clientSecret;
+    private String authorization;
+    private ConfidentialClientApplication confidentialClientApplication;
 
-    KeyVaultCredential(String clientId) throws SQLServerException {
+    /**
+     * Creates a KeyVaultCredential with the given identity client options.
+     *
+     * @param clientId
+     *        the client ID of the application
+     * @param clientSecret
+     *        the secret value of the AAD application.
+     */
+    KeyVaultCredential(String clientId, String clientSecret) {
+        Objects.requireNonNull(clientSecret, "'clientSecret' cannot be null.");
+        Objects.requireNonNull(clientSecret, "'clientId' cannot be null.");
         this.clientId = clientId;
+        this.clientSecret = clientSecret;
     }
 
-    KeyVaultCredential() {}
-
-    KeyVaultCredential(String clientId, String clientKey) {
-        this.clientId = clientId;
-        this.clientKey = clientKey;
+    @Override
+    public Mono<AccessToken> getToken(TokenRequestContext request) {
+        return authenticateWithConfidentialClientCache(request).onErrorResume(t -> Mono.empty())
+                .switchIfEmpty(Mono.defer(() -> authenticateWithConfidentialClient(request)));
     }
 
-    KeyVaultCredential(SQLServerKeyVaultAuthenticationCallback authenticationCallback) {
-        this.authenticationCallback = authenticationCallback;
+    public KeyVaultCredential setAuthorization(String authorization) {
+        if (null != this.authorization && this.authorization.equals(authorization)) {
+            return this;
+        }
+        this.authorization = authorization;
+        confidentialClientApplication = getConfidentialClientApplication();
+        return this;
     }
 
-    public String doAuthenticate(String authorization, String resource, String scope) {
-        String accessToken = null;
-        if (null == authenticationCallback) {
-            if (null == clientKey) {
-                try {
-                    SqlFedAuthToken token = SQLServerSecurityUtility.getMSIAuthToken(resource, clientId);
-                    accessToken = (null != token) ? token.accessToken : null;
-                } catch (Exception e) {
-                    throw new RuntimeException(e);
-                }
-            } else {
-                AuthenticationResult token = getAccessTokenFromClientCredentials(authorization, resource, clientId,
-                        clientKey);
-                accessToken = token.getAccessToken();
-            }
+    private ConfidentialClientApplication getConfidentialClientApplication() {
+        if (null == clientId) {
+            throw logger.logExceptionAsError(new IllegalArgumentException(
```

Comment:
This is not how exception are handled in the driver, @lilgreenbird please make sure to follow the guidelines.